rclrs/src/timer.rs

```diff
@@ -0,0 +1,416 @@
+    where
+        F: FnMut(&mut Timer) + 'static + Send + Sync,
+    {
+        let callback = Arc::new(Mutex::new(callback));
+
+        // SAFETY: Getting a zero-initialized value is always safe.
+        let mut rcl_timer = unsafe { rcl_get_zero_initialized_timer() };
+
+        let clock_clone = clock.rcl_clock.clone();
+        let mut rcl_clock = clock_clone.lock().unwrap();
+
+        let context_handle_clone = context_handle.clone();
+        let mut rcl_context = context_handle_clone.rcl_context.lock().unwrap();
+
+        // core::time::Duration will always be >= 0, so no need to check for negatives.
+        let period_nanos = i64::try_from(period.as_nanos()).unwrap_or(i64::MAX);
+
+        // SAFETY: No preconditions for this function.
+        let allocator = unsafe { rcutils_get_default_allocator() };
+        {
+            let _lifecycle_lock = ENTITY_LIFECYCLE_MUTEX.lock().unwrap();
+            unsafe {
+                // SAFETY:
+                // * The rcl_timer is zero-initialized as mandated by this function.
+                // * The rcl_clock is kept alive by the Clock within TimerHandle because it is
+                //   a dependency of the timer.
+                // * The rcl_context is kept alive by the ContextHandle within TimerHandle because
+                //   it is a dependency of the timer.
+                // * The period is copied into this function so it can be dropped afterwards.
+                // * The callback is None / nullptr so doesn't need to be kept alive.
+                // * The entity lifecycle mutex is locked to protect against the risk of global
+                //   variables in the rmw implementation being unsafely modified during initialization.
+                rcl_timer_init(
+                    &mut rcl_timer,
+                    &mut *rcl_clock,
+                    &mut *rcl_context,
+                    period_nanos,
+                    None,
+                    allocator,
+                )
+                .ok()?;
+            }
+        }
+
+        Ok(Self {
+            callback,
+            handle: TimerHandle {
+                rcl_timer: Mutex::new(rcl_timer),
+                _clock: clock,
+                _context_handle: context_handle,
+                in_use_by_wait_set: Arc::new(AtomicBool::new(false)),
+            },
+        })
+    }
+
+    /// Calculates if the timer is ready to be called.
+    /// Returns true if the timer is due or past due to be called.
+    /// Returns false if the timer is not yet due or has been canceled.
+    pub fn is_ready(&self) -> bool {
+        let mut timer = self.handle.lock();
+        let mut is_ready = false;
+        // SAFETY:
+        // * The timer is initialized, which is guaranteed by the constructor.
+        // * The is_ready pointer is allocated on the stack and is valid for the duration of this function.
+        let ret = unsafe { rcl_timer_is_ready(&mut *timer, &mut is_ready) };
```

Comment:
As Per: https://github.com/geoff-imdex/ros2_rust/actions/runs/11564633206/job/32190192780#step:11:395

```suggestion
        let ret = unsafe { rcl_timer_is_ready(& *timer, &mut is_ready) };
```

Reply:
Good catch, fixed.